irulan.1
.Dd August 10 2019
.Dt IRULAN 1
.nh
.Os
.Sh NAME
.Nm irulan
.Nd ssh host key management tool
.Sh SYNOPSIS
.Bk -words
.Nm
.Op Fl P Ar program
.Op Fl n
.Ar command
.Op Ar cmd-args ..
.Ek
.Sh DESCRIPTION
.Nm
interacts with the Irulan database of the Web::Irulan Perl module and
provides various commands related to the management of ssh host keys. It
runs on the same host that the Irulan database is hosted on.
.Ss Options
These may appear before or after the
.Ar command .
.Bl -tag -width Ds
.It Fl P Ar program
Program to exec when a
.Ar command
that alters the database is run. The
.Ar command
name is passed as the first argument to the
.Ar program
and can be used to determine whether a subsequent
.Nm
.Cm ssh_known_hosts
command can use epoch time caching or not: yes for
.Cm addhost
and no for 
.Cm rmhost
or
.Cm rmsystem .
.It Fl n 
Do not perform hostname to IP address conversions. Not advised with the
.Cm ssh_known_hosts
command if there are hostnames in the database, as 
.Xr ssh 1
may then add entries for IP addresses to
.Pa known_hosts
files.
.El
.Ss Commands
A 
.Ar command
must be provided. Port 22 is assumed for various commands if not
provided.
.Bl -tag -width Ds
.It Cm addhost Ar sysid Ar host Op Ar port
Associates a system ID with a particular hostname. Use the
.Cm list
or
.Cm unhosted
commands to see existing or pending records to add.
Executes the
.Fl P Ar program
with
.Ar addhost
as the first argument.
.It Cm audit Op Ar host Op Ar port
Audits the given host or lacking that all hosts by connecting to each
host via the
.Dv IRULAN_AUDIT_CMD
that must be set; this command must support three arguments:
.Bl -enum
.It
.Ar known-host-file
.It
.Ar host
.It
.Op Ar port
.El
.It Cm cat Op Ar file|- ..
Parses public keys from the given files or stdin and associates all the
keys parsed with a new system.
.It Cm keyscan Ar host Op Ar port
Scans a system using
.Xr ssh-keyscan 1
and parses the public keys found into a new system entry. Pipe
instead to the
.Nm
.Cm cat
command if custom options or filtering need be done.
.It Cm list
Lists the records in the database.
.It Cm rmhost Ar host Op Ar port
Removes a host mapping for a particular system. Use this if a CNAME
now points to a new host, or if a raw IP address in the database
has changed.
Executes the
.Fl P Ar program
with
.Ar rmhost
as the first argument.
.It Cm rmsystem Ar sysid
Removes a system and all associated records from the database, such as
when a host is completely retired.
Executes the
.Fl P Ar program
with
.Ar rmsystem
as the first argument.
.It Cm ssh_known_hosts Op Ar epoch
Generates ssh known hosts data to standard output. Hostnames will be
resolved to determine their IP address(es) unless the
.Fl n
flag is given. If a unix epoch time value is provided, the data will be
generated only if any record is more recent than the given epoch value.
.It Cm unhosted
Lists systems that have no associated host. Use
.Cm addhost
to make such an association.
.El
.Sh ENVIRONMENT
.Bl -tag -width Ds
.It Dv IRULAN_AUDIT_CMD
Command used by the
.Cm audit
mode to check host key entries in the database.
.It Dv IRULAN_DB_PATH
Specifies the path to the Irulan database. This path should agree with
what the Web::Irulan web application uses, if that is being used.
.El
.Sh EXIT STATUS
.Ex -std
.Pp
The 
.Fl P Ar program
flag replaces
.Nm
with the given
.Ar program
so when that is used the exit status will be of that tool.
.Sh EXAMPLES
Assuming a client has uploaded public keys and an admin has been
notified about this somehow, the system associated with the client can
be found and then a hostname record added for that client;
.Pp
.Dl $ Ic irulan unhosted
.Dl ...
.Dl $ Ic irulan addhost 123 foo.example.edu
.Pp
would display what system records have no corresponding hostname, and
the second command would associate that specific hostname with system id
123. Hosts must manually be added.
.Pp
A script such as
.Pp
.Dl #!/bin/sh
.Dl set -e
.Dl cd \&"$HOME\&"/.ssh
.Dl irulan ssh_known_hosts > kh.irulan
.Dl cat kh.* > known_hosts
.Pp
could then be used to export the known hosts and build
.Pa ~/.ssh/known_hosts
for a user. Another option would be to store all the ssh host keys for
an organization and export from
.Nm
into 
.Pa /etc/ssh/ssh_known_hosts
on all the hosts.
.Pp
The known hosts for this system might be collected with
.Cm cat :
.Pp
.Dl $ Ic irulan cat /etc/ssh/ssh_host_*.pub 
.Pp
Loopback or private network virts running on strange ports without DNS
can also be added
.Pp
.Dl $ Ic irulan -n keyscan 192.0.2.42 2222
.Pp
though these may not suit a globally deployed
.Pa ssh_known_hosts
file if they can only be reached by a subset of the systems.
.Sh SEE ALSO
.Xr ssh-hkaudit 1 ,
.Xr ssh-keyscan 1
.Pp
.Em https://github.com/thrig/web_irulan
.Sh AUTHOR
.An Jeremy Mates


lib/Web/Irulan.pm
# -*- Perl -*-

package Web::Irulan;
use Mojo::Base 'Mojolicious';

use Web::Irulan::Model::Hostkeys;
use Mojo::SQLite;

sub startup {
    my ($self) = @_;

    my $config = $self->plugin('Config');
    $self->secrets($config->{secrets});

    $self->helper(
        sqlite => sub {
            state $sql =
              Mojo::SQLite->new->from_filename($ENV{IRULAN_DB_PATH} // 'irulan.db');
        }
    );
    $self->helper(
        hostkeys => sub {
            state $hkeys = Web::Irulan::Model::Hostkeys->new(sqlite => $_[0]->sqlite);
        }
    );

    my $path = $self->home->child(qw(schema irulan.sql));
    $self->sqlite->auto_migrate(1)->migrations->name('irulan')->from_file($path);

    # app may be running under some not-/ path that it will need to
    # know about
    if (my $path = $ENV{MOJO_REVERSE_PROXY}) {
        my @path_parts = grep /\S/, split m{/}, $path;
        $self->hook(
            before_dispatch => sub {
                my ($c)  = @_;
                my $url  = $c->req->url;
                my $base = $url->base;
                push @{ $base->path }, @path_parts;
                $base->path->trailing_slash(1);
                $url->path->leading_slash(0);
            }
        );
    }

    my $r = $self->routes;
    $r->get('/')->to(
        cb => sub {
            my ($c) = @_;
            $c->render(text => 'Irulan');
        }
    );

    my $ssh = $r->under(
        '/ssh' => sub {
            my ($c) = @_;
            my $token = $c->req->headers->header('X-Irulan-Auth') // '';
            return 1 if $token eq $config->{authkey};
            $c->render(json => { error => 'unauthorized' }, status => 401);
            return undef;
        }
    );
    $ssh->post('/hostkeys')->to('ssh#upload');
}

1;
